README.md
# BTCrypto

![OS](https://img.shields.io/badge/OS-JRE-darkgreen)
![licence](https://img.shields.io/badge/language-Java-brightgreen.svg?style=flat-square)
![license](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)

Implements cryptographic algorithms on Java, including the official [CuaimaCrypt](#what-is-cuaimaCrypt) source code.

This project is provided "AS IS", no warranty, no responsibilities, no more documentation that the one included in this
repository.

## What is CuaimaCrypt

CuaimaCrypt library was born on May 28, 2007.

The first decision was what kind of algorithm would be, one of block coding or pseudo-random sequences coding bits; 
after much research and testing we integrate both models into one, developing "Shift Codec", which allows a non-cyclic
code with pseudo-random sequence.

These Shift Codecs are grouped in Rake Codec, which can process simultaneously 128 bits and these are grouped in
sequence of bits by scramblers with four possible operations. The Shift Codecs are chained with other Shift Codecs,
which makes his states dependant of the data entered into the process; additionally a sequence of Shift Codec's hopping
states.

This coding structure allows to extend the security of the algorithm to the particular needs of everyone just adding
more blocks.

All the initial parameters of the algorithm, seeds, scramblers between blocks, seeds hopping sequence, Shift Codec
chains, all of them are initialized by the key who define the Lorenz attractors' states.

The equation that define the possible combinations of CuaimaCrypt’s configuration is:

CuaimaCrypt-Equation

![CuaimaCrypt Equation](./img/cuaimacrypt-equation.png)

Where N is the number of blocks in the algorithm.

| Number of Blocks N | CuaimaCrypt’s Configuration Combinations | Equivalent Key Bits Length |
|--------------------|------------------------------------------|----------------------------|
| 2                  | 3.6855x10 ^ 165                          | 550                        |
| 4                  | 5.4648x10 ^ 324                          | 1078                       |
| 6                  | 7.1220x10 ^ 484                          | 1610                       |
| 9                  | 2.0384x10 ^ 726                          | 2412                       |
| 16                 | 5.6466x10 ^ 1293                         | 4297                       |
| 32                 | 2.0153x10 ^ 2603                         | 8647                       |
| 64                 | 1.8187x10 ^ 5250                         | 17440                      |
| 128                | 8.2660x10 ^ 10600                        | 35215                      |
| 256                | 5.9661x10 ^ 21416                        | 71144                      |
| 512                | 4.2701x10 ^ 43278                        | 143768                     |
| 1024               | 4.6637x10 ^ 87463                        | 290548                     |

In perspective at the observable universe we have in a gross underestimation of 1x10^29 stars according to astronomers'
estimates. If we compare these values with known symmetric key algorithms we can see:

- AES: 256 bits
- Blowfish: 448 bits
- Serpent: 256 bits
- Twofish: 256 bits
- GOST: 256 bits
- REDOC: 160 bits
- IDEA: 128 bits
- MMB: 128 bits
- SAFER K-128: 128 bits
- NewDES: 120 bits
- 3-WAY: 96 bits
- CRAB: 80 bits
- SKIPJACK: 80 bits (Clasificated algorithm by NSA)
- CAST: 64 bits
- Madryga: 64 bits
- FEAL: 64 bits
- LOKI: 64 bits
- SAFER K-64: 64 bits
- DES: 56 bits

As we can see with 6 blocks; CuaimaCrypt has 1610 bits, which is about 3.5 times the level of security offered by Blowfish.

Another CuaimaCrypt's feature is that the security level is not determined by the length of the password, in the extreme
case that key is just one letter, all the initial values are generated by the Lorenz attractors, one key of less than
9 chars is not recommended as it would be extremely vulnerable to a brute force attack against the key itself, so the
security of the key is the user's responsibility.

At algorithm tests, medium files (1 MB) in size, it has a completely flat distribution of characters, with a variation
of only 0.05% maximum over all the ASCII table.


## Dependencies

This library depends on [BTUtils](https://github.com/BolivarTech/BTUtils) library.

## Credits

- [Julian Bolivar](https://www.linkedin.com/in/jbolivarg/)

## License

Copyright © [BolivarTech](https://www.bolivartech.com) 2023 All Rights Reserved.

BtCrypto and CuaimaCrypt © 2023 by [Julian Bolivar](https://www.bolivartech.com) is licensed under [Attribution-ShareAlike 4.0 International](https://creativecommons.org/licenses/by-sa/4.0/legalcode)

Please see [License File](LICENSE.md) for more information.

CHANGELOG.md
# CHANGELOG.md

---

## 1.0.0 (2023-06-25)

Security:

- No security risk associated

Features:

- Java General Use Cryptographic Library

Fix:

- No fixes were necessary

Bugfixes:

- No bugs found


src/com/bolivartech/crypto/digest/MD2.java
package com.bolivartech.crypto.digest;

/**
 * <p>An implementation of the MD2 message digest algorithm.</p>
 *
 * <p>MD2 is not widely used. Unless it is needed for compatibility with
 * existing systems, it is not recommended for use in new applications.</p>
 *
 * <p>References:</p>
 *
 * <ol>
 *    <li>The <a href="http://www.ietf.org/rfc/rfc1319.txt">MD2</a>
 *    Message-Digest Algorithm.<br>
 *    B. Kaliski.</li>
 *    <li>The <a href="http://www.rfc-editor.org/errata.html">RFC ERRATA PAGE</a>
 *    under section RFC 1319.</li>
 * </ol>
 *
 * @author Julian Bolivar
 * @since 2010 - December 11, 2010
 * @version $Revision: 1.0 $
 */
public class MD2 extends BaseHash {

   // Constants and variables
   // -------------------------------------------------------------------------

   /** An MD2 message digest is always 128-bits long, or 16 bytes. */
   private static final int DIGEST_LENGTH = 16;

   /** The MD2 algorithm operates on 128-bit blocks, or 16 bytes. */
   private static final int BLOCK_LENGTH = 16;

   /** 256 byte "random" permutation of the digits of pi. */
   private static final byte[] PI = {
      41, 46, 67, -55, -94, -40, 124, 1, 61, 54, 84, -95, -20, -16, 6,
      19, 98, -89, 5, -13, -64, -57, 115, -116, -104, -109, 43, -39,
      -68, 76, -126, -54, 30, -101, 87, 60, -3, -44, -32, 22, 103, 66,
      111, 24, -118, 23, -27, 18, -66, 78, -60, -42, -38, -98, -34, 73,
      -96, -5, -11, -114, -69, 47, -18, 122, -87, 104, 121, -111, 21,
      -78, 7, 63, -108, -62, 16, -119, 11, 34, 95, 33, -128, 127, 93,
      -102, 90, -112, 50, 39, 53, 62, -52, -25, -65, -9, -105, 3, -1,
      25, 48, -77, 72, -91, -75, -47, -41, 94, -110, 42, -84, 86, -86,
      -58, 79, -72, 56, -46, -106, -92, 125, -74, 118, -4, 107, -30,
      -100, 116, 4, -15, 69, -99, 112, 89, 100, 113, -121, 32, -122,
      91, -49, 101, -26, 45, -88, 2, 27, 96, 37, -83, -82, -80, -71,
      -10, 28, 70, 97, 105, 52, 64, 126, 15, 85, 71, -93, 35, -35, 81,
      -81, 58, -61, 92, -7, -50, -70, -59, -22, 38, 44, 83, 13, 110,
      -123, 40, -124, 9, -45, -33, -51, -12, 65, -127, 77, 82, 106,
      -36, 55, -56, 108, -63, -85, -6, 36, -31, 123, 8, 12, -67, -79,
      74, 120, -120, -107, -117, -29, 99, -24, 109, -23, -53, -43, -2,
      59, 0, 29, 57, -14, -17, -73, 14, 102, 88, -48, -28, -90, 119,
      114, -8, -21, 117, 75, 10, 49, 68, 80, -76, -113, -19, 31, 26,
      -37, -103, -115, 51, -97, 17, -125, 20
   };

   /** The output of this message digest when no data has been input. */
   private static final String DIGEST0 = "8350E5A3E24C153DF2275C9F80692773";

   /** caches the result of the correctness test, once executed. */
   private static Boolean valid;

   /** The checksum computed so far. */
   private byte[] checksum;

   /**
    * Work array needed by encrypt method. First <code>BLOCK_LENGTH</code> bytes
    * are also used to store the running digest.
    */
   private byte[] work;

   // Constructor(s)
   // -------------------------------------------------------------------------

   /** Creates a new MD2 digest ready for use. */
   public MD2() {
      super(Registry.MD2_HASH, DIGEST_LENGTH, BLOCK_LENGTH);
   }

   /**
    * <p>Private constructor used for cloning.</p>
    *
    * @param md2 the instance to clone.
    */
   private MD2(MD2 md2) {
      this();

      // superclass field
      this.count = md2.count;
      this.buffer = (byte[]) md2.buffer.clone();

      // private field
      this.checksum = (byte[]) md2.checksum.clone();
      this.work = (byte[]) md2.work.clone();
   }

   // Class methods
   // -------------------------------------------------------------------------

   // Instance methods
   // -------------------------------------------------------------------------

   // java.lang.Cloneable interface implementation ----------------------------

   @Override
   public Object clone() {
      return new MD2(this);
   }

   // Implementation of abstract methods in BaseHash --------------------------

   @Override
   protected byte[] getResult() {
      byte[] result = new byte[DIGEST_LENGTH];

      // Encrypt checksum as last block.
      encryptBlock(checksum, 0);
      System.arraycopy(work, 0, result, 0, BLOCK_LENGTH);
/*      for (int i = 0; i < BLOCK_LENGTH; i++) {
         result[i] = work[i];
      } */
      return result;
   }

   @Override
   protected void resetContext() {
      checksum = new byte[BLOCK_LENGTH];
      work = new byte[BLOCK_LENGTH*3];
   }

   @Override
   public boolean selfTest() {
      MD2 Test;
       
      if (valid == null) {
         Test = new MD2();
         valid = DIGEST0.equals(StringUtils.toString(Test.digest()));
      }
      return valid.booleanValue();
   }

   /**
    * <p>Generates an array of padding bytes. The padding is defined as
    * <code>i</code> bytes of value <code>i</code>, where <code>i</code> is the
    * number of bytes to fill the last block of the message to
    * <code>BLOCK_LENGTH</code> bytes (or <code>BLOCK_LENGTH</code> bytes when
    * the last block was completely full).</p>
    *
    * @return the bytes to pad the remaining bytes in the buffer before
    * completing a hash operation.
    */
   @Override
   protected byte[] padBuffer() {
      int length = BLOCK_LENGTH - (int)(count % BLOCK_LENGTH);
      if (length == 0) {
         length = BLOCK_LENGTH;
      }
      byte[] pad = new byte[length];
      for (int i = 0; i < length; i++) {
         pad[i] = (byte) length;
      }
      return pad;
   }

   /**
    * <p>Adds <code>BLOCK_LENGTH</code> bytes to the running digest.</p>
    *
    * @param in the byte array to take the <code>BLOCK_LENGTH</code> bytes from.
    * @param off the offset to start from in the given byte array.
    */
   @Override
   protected void transform(byte[] in, int off) {
      // encryptBlock(in, off);
      // updateCheckSum(in, off);
      updateCheckSumAndEncryptBlock(in, off);
   }

   // Private instance methods ------------------------------------------------

   /**
    * Updates the checksum with the <code>BLOCK_LENGTH</code> bytes from the
    * given array starting at <code>off</code>.
    */
/*
   private void updateCheckSum(byte[] in, int off) {
      byte l = checksum[BLOCK_LENGTH-1];
      for (int i = 0; i < BLOCK_LENGTH; i++) {
         byte b = in[off+i];
//         l = (byte)((checksum[i] & 0xFF) ^ (PI[((b & 0xFF) ^ (l & 0xFF))] & 0xFF));
         l = (byte)(checksum[i] ^ PI[(b ^ l) & 0xFF]);
         checksum[i] = l;
      }
   }
*/
   /**
    * Adds a new block (<code>BLOCK_LENGTH</code> bytes) to the running digest
    * from the given byte array starting from the given offset.
    */
   private void encryptBlock(byte[] in, int off) {
      for (int i = 0; i < BLOCK_LENGTH; i++) {
         byte b = in[off+i];
         work[BLOCK_LENGTH+i] = b;
         work[BLOCK_LENGTH*2+i] = (byte)(work[i] ^ b);
      }

      byte t = 0;
      for (int i = 0; i < 18; i++) {
         for (int j = 0; j < 3 * BLOCK_LENGTH; j++) {
//            t = (byte)((work[j] & 0xFF) ^ (PI[t & 0xFF] & 0xFF));
            t = (byte)(work[j] ^ PI[t & 0xFF]);
            work[j] = t;
         }
//         t = (byte)((t + i) & 0xFF);
         t = (byte)(t + i);
      }
   }

   /**
    * Optimized method that combines a checksum update and encrypt of a block.
    */
   private void updateCheckSumAndEncryptBlock(byte[] in, int off) {
      byte l = checksum[BLOCK_LENGTH-1];
      for (int i = 0; i < BLOCK_LENGTH; i++) {
         byte b = in[off+i];
         work[BLOCK_LENGTH+i] = b;
//         work[BLOCK_LENGTH*2+i] = (byte)((work[i] & 0xFF) ^ (b & 0xFF));
         work[BLOCK_LENGTH*2+i] = (byte)(work[i] ^ b);
//         l = (byte)((checksum[i] & 0xFF) ^ (PI[((b & 0xFF) ^ (l & 0xFF))] & 0xFF));
         l = (byte)(checksum[i] ^ PI[(b ^ l) & 0xFF]);
        checksum[i] = l;
      }

      byte t = 0;
      for (int i = 0; i < 18; i++) {
         for (int j = 0; j < 3 * BLOCK_LENGTH; j++) {
//            t = (byte)((work[j] & 0xFF) ^ (PI[t & 0xFF] & 0xFF));
            t = (byte)(work[j] ^ PI[t & 0xFF]);
            work[j] = t;
         }
//         t = (byte)((t + i) & 0xFF);
         t = (byte)(t + i);
      }
   }

}
